apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
+                    // Get all ack-ids, used to extend leases as required.
+                    var msgIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                    // Send an initial "lease-extension"; which starts the server timer.
+                    HandleExtendLease(msgIds);
+                    // Asynchonously start message processing. Handles flow, and calls the user-supplied message handler.
+                    // Uses Task.Run(), so not to clog up this "master" thread with per-message processing.
+                    Task messagesTask = _taskHelper.Run(() => ProcessPullMessagesAsync(msgs, msgIds));
+                    // Once all received messages have been queued for processing, read the stream for more messages.
+                    Add(messagesTask, Next(true, () => HandlePullMoveNext(null)));
+                }
+                else
+                {
+                    StopStreamingPull();
+                    // Always a short pause on server disconnect.
+                    _pullBackoff = TimeSpan.FromSeconds(0.5);
+                    StartStreamingPull();
                 }
             }
 
-            private void QueueLeaseExtensions(IEnumerable<string> ackIds)
+            private async Task ProcessPullMessagesAsync(List<ReceivedMessage> msgs, HashSet<string> msgIds)
             {
-                lock (_qLock)
+                // Running async. Common data needs locking
+                for (int msgIndex = 0; msgIndex < msgs.Count; msgIndex++)
                 {
-                    _extendQueue.EnqueueAll(ackIds);
+                    _softStopCts.Token.ThrowIfCancellationRequested();
+                    var msg = msgs[msgIndex];
+                    msgs[msgIndex] = null;
+                    await _taskHelper.ConfigureAwait(_flow.Process(msg.CalculateSize(), async () =>
+                    {
+                            // Running async. Common data needs locking
+                            lock (_lock)
+                        {
+                            _softStopCts.Token.ThrowIfCancellationRequested();
+                            _userHandlerInFlight += 1;
+                        }
+                            // Call user message handler
+                            var reply = await _taskHelper.ConfigureAwaitHideErrors(() => _handlerAsync(msg.Message, _hardStopCts.Token), Reply.Nack);
+                            // Lock msgsIds, this is accessed concurrently here and in HandleExtendLease().
+                            lock (msgIds)
+                        {
+                            msgIds.Remove(msg.AckId);
+                        }
+                            // Lock ack/nack-queues, this is accessed concurrently here and in "master" thread.
+                            lock (_lock)
+                        {
+                            _userHandlerInFlight -= 1;
+                            (reply == Reply.Ack ? _ackQueue : _nackQueue).Enqueue(msg.AckId);
```

Comment:
Extract the conditional expression to a local variable?

Reply:
OK, done.